### PR TITLE
Add warning on long Auto Type sequences

### LINKED ISSUE #216: Delays > 10000 are quietly discarded
The "safety-checks" in https://github.com/keepassxreboot/keepassxc/blob/develop/src/autotype/AutoType.cpp#L384 and https://github.com/keepassxreboot/keepassxc/blob/develop/src/autotype/AutoType.cpp#L388 don't really make sense to me.

## Expected Behavior
If I enter a autotype sequence with '{delay 10001}' I expect it to wait for 10001ms.
If I enter '{left 101}', I expect it to type left 101 times.

Alternatively there should be some sort of warning that these values are invalid.

## Current Behavior
Delays greater than 10000 and repetitions greater than 100 are silently ignored.

## Possible Solution

Remove the checks or show a warning when these values are entered.
Alternatively go the same way as in https://github.com/keepassxreboot/keepassxc/blob/develop/src/autotype/AutoType.cpp#L370 and enforce bounds for the values. Although I think this isn't a good solution either as it quietly disobeys the users instructions.

### PR DESCRIPTION
**Pull request by TheZ3ro**
Replace #552 all credits to @marcbone

Auto Type now shows a warning when you try to  repeat something too often.
Also you can now repeat your password and username

I'm new to all this open source, transifex,qt and c++ stuff. So dont spare with criticism but be nice.
I didnt know how to get the message-strings away from the code and had some problems with the tests, so it would be better if somebody looks over it. Sorry for that.

## Description

Large delays and large repetitions arent ignored anymore, instead you get a Messagebox asking you if you want to continue or abort.

## Motivation and context

Fixes #216 

## How has this been tested?

I changed autoType command to repeat it 101 times and executed the autotype command.

Problem is, if you have multiple commands which are repeated very often (ex. {Password 101} {Username 101}) you get multiple MessageBoxes. I could have changed that, but I didnt find a way to do so without bad code or bigger changes.

Tested on Ubuntu 16.04

## Screenshots (if appropriate):
bildschirmfoto vom 2017-05-08 16-28-48
## Types of changes

- ✅ Bug fix (non-breaking change which fixes an issue)

## Checklist:

- ✅ I have read the **CONTRIBUTING** document.
- ✅ My code follows the code style of this project. 
-  All new and existing tests passed. GUI-Test failed, bui it also failed without my changes so I guess its ok
- ✅ I have compiled and verified my code with `-DWITH_ASAN=ON`. (I have testesd coverage and directly the autotype features, diddnt get warnings on the console, but I really dont know what I should do there exactly

### CODE REVIEW

File: src/autotype/AutoType.cpp [hunk 1]
```diff
@@ -658,3 +660,76 @@ bool AutoType::windowMatchesUrl(const QString& windowTitle, const QString& resol
...
+{
+    if (!AutoType::checkSyntax(entry->effectiveAutoTypeSequence())) {
+        QMessageBox messageBox;
+        messageBox.critical(0, "AutoType", tr("The Syntax of your AutoType statement is incorrect!"));
+        return;
+    }
+    else if (AutoType::checkHighDelay(entry->effectiveAutoTypeSequence())) {
+        QMessageBox::StandardButton reply;
+        reply = QMessageBox::question(0,
+                                      "AutoType",
```
Comment (weslly) [inline src/autotype/AutoType.cpp]: Shouldn't the message title be translated with `tr()` too?

File: src/autotype/AutoType.cpp [hunk 2]
```diff
@@ -596,7 +597,7 @@ QString AutoType::autoTypeSequence(const Entry* entry, const QString& windowTitl
         sequence = entry->defaultAutoTypeSequence();
     }
 
-    const Group* group = entry->group();
+    const Group *group = entry->group();
```
Comment (weslly) [inline src/autotype/AutoType.cpp]: `Group* group`

File: src/autotype/AutoType.h [hunk 3]
```diff
@@ -34,13 +34,19 @@ class AutoType : public QObject
...
-    void performAutoType(const Entry* entry, QWidget* hideWindow = nullptr,
+    void _performAutoType(const Entry* entry, QWidget* hideWindow = nullptr,
                          const QString& customSequence = QString(), WId window = 0);
     bool registerGlobalShortcut(Qt::Key key, Qt::KeyboardModifiers modifiers);
     void unregisterGlobalShortcut();
     int callEventFilter(void* event);
+    static bool checkSyntax(const QString &string);
+    static bool checkHighRepetition(const QString &string);
+    static bool checkHighDelay(const QString &string);
+    void performAutoType(const Entry *entry, QWidget *hideWindow = nullptr,
```
Comment (weslly) [inline src/autotype/AutoType.h]: Pointers on left

**Comment by TheZ3ro** [comment]
@weslly fixed

File: src/autotype/AutoType.h (hunk 3 above)
Comment (louib) [reply src/autotype/AutoType.h]: @TheZ3ro this was not addressed

Comment (TheZ3ro) [reply src/autotype/AutoType.h]: done :wink:

File: src/gui/entry/EditEntryWidget.cpp [hunk 4]
```diff
@@ -499,7 +500,40 @@ void EditEntryWidget::updateEntryData(Entry* entry) const
         entry->setDefaultAutoTypeSequence(QString());
     }
     else {
-        entry->setDefaultAutoTypeSequence(m_autoTypeUi->sequenceEdit->text());
+        if (!AutoType::checkSyntax(m_autoTypeUi->sequenceEdit->text())) {
+            //handle wrong syntax
+            QMessageBox messageBox;
+            messageBox.critical(0,
+                                "AutoType",
+                                tr("The Syntax of your AutoType statement is incorrect! It won't be saved!"));
```
Comment (phoerious) [inline src/gui/entry/EditEntryWidget.cpp]: Auto-Type

File: src/gui/entry/EditEntryWidget.cpp [hunk 5]
```diff
@@ -499,7 +500,40 @@ void EditEntryWidget::updateEntryData(Entry* entry) const
...
+
+        }
+        else if (AutoType::checkHighDelay(m_autoTypeUi->sequenceEdit->text())) {
```
Comment (phoerious) [inline src/gui/entry/EditEntryWidget.cpp]: No newline before else (more instances of this in this file).

File: src/gui/entry/EditEntryWidget.cpp [hunk 6]
```diff
@@ -499,7 +500,40 @@ void EditEntryWidget::updateEntryData(Entry* entry) const
...
+            //handle too long delay
+            QMessageBox::StandardButton reply;
+            reply = QMessageBox::question(0,
+                                          "AutoType",
+                                          tr("This AutoType command contains a very long delay. Do you really want to save it?"));
```
Comment (phoerious) [inline src/gui/entry/EditEntryWidget.cpp]: Auto-Type

File: src/gui/entry/EditEntryWidget.cpp [hunk 7]
```diff
@@ -499,7 +500,40 @@ void EditEntryWidget::updateEntryData(Entry* entry) const
...
+            if (reply == QMessageBox::Yes) {
+                entry->setDefaultAutoTypeSequence(m_autoTypeUi->sequenceEdit->text());
+            }
+        }
+        else if (AutoType::checkHighRepetition(m_autoTypeUi->sequenceEdit->text())) {
+            //handle too much repetition
+            QMessageBox::StandardButton reply;
+            reply = QMessageBox::question(0,
+                                          "AutoType",
+                                          tr("This AutoType command contains arguments which are repeated very often. Do you really want to save it?"));
```
Comment (phoerious) [inline src/gui/entry/EditEntryWidget.cpp]: Auto-Type

File: src/gui/entry/EditEntryWidget.cpp (hunk 4 above)
Comment (weslly) [reply src/gui/entry/EditEntryWidget.cpp]: `tr()`

File: src/gui/entry/EditEntryWidget.cpp (hunk 6 above)
Comment (weslly) [reply src/gui/entry/EditEntryWidget.cpp]: `tr()`

File: src/gui/entry/EditEntryWidget.cpp (hunk 7 above)
Comment (weslly) [reply src/gui/entry/EditEntryWidget.cpp]: `tr()`

File: src/gui/entry/EditEntryWidget.cpp (hunk 6 above)
Comment (phoerious) [reply src/gui/entry/EditEntryWidget.cpp]: Good point. Totally missed that.

**Comment by TheZ3ro** [comment]
fixed and rebased

File: src/autotype/AutoType.h [hunk 8]
```diff
@@ -36,13 +36,23 @@ class AutoType : public QObject
 
 public:
     QStringList windowTitles();
-    void performAutoType(const Entry* entry, QWidget* hideWindow = nullptr,
-                         const QString& customSequence = QString(), WId window = 0);
+    void _performAutoType(const Entry* entry,
```
Comment (phoerious) [inline src/autotype/AutoType.h]: I don't like this naming. Name it doPerformanceAutoType or whatever, but don't start with an underscore.

Comment (TheZ3ro) [reply src/autotype/AutoType.h]: fixed

**Comment by TheZ3ro** [comment]
@phoerious fixed and rebased

**Comment by TheZ3ro** [comment]
fixed conflicts and rebased. Can someone review this? @keepassxreboot/core-developers

File: src/gui/entry/EditEntryWidget.cpp [hunk 9]
```diff
@@ -179,24 +184,25 @@ void EditEntryWidget::setupAutoType()
...
-    connect(m_autoTypeUi->assocView->selectionModel(), SIGNAL(currentRowChanged(QModelIndex,QModelIndex)),
+    connect(m_autoTypeUi->assocView->selectionModel(),
+            SIGNAL(currentRowChanged(QModelIndex, QModelIndex)),
             SLOT(updateAutoTypeEnabled()));
     connect(m_autoTypeAssocModel, SIGNAL(modelReset()), SLOT(updateAutoTypeEnabled()));
-    connect(m_autoTypeUi->assocView->selectionModel(), SIGNAL(currentRowChanged(QModelIndex,QModelIndex)),
+    connect(m_autoTypeUi->assocView->selectionModel(),
+            SIGNAL(currentRowChanged(QModelIndex, QModelIndex)),
             SLOT(loadCurrentAssoc(QModelIndex)));
     connect(m_autoTypeAssocModel, SIGNAL(modelReset()), SLOT(clearCurrentAssoc()));
```
Comment (phoerious) [inline src/gui/entry/EditEntryWidget.cpp]: Are all these reformats really necessary? They kind of convolute the diff for no real gain. It's also not quite comprehensible when you use a single line and when you split an expression.

File: src/gui/entry/EditEntryWidget.cpp [hunk 10]
```diff
@@ -763,20 +757,43 @@ void EditEntryWidget::updateEntryData(Entry* entry) const
...
         entry->setIcon(iconStruct.uuid);
     }
 
     entry->setAutoTypeEnabled(m_autoTypeUi->enableButton->isChecked());
     if (m_autoTypeUi->inheritSequenceButton->isChecked()) {
         entry->setDefaultAutoTypeSequence(QString());
-    }
-    else {
-        entry->setDefaultAutoTypeSequence(m_autoTypeUi->sequenceEdit->text());
+    } else {
```
Comment (phoerious) [inline src/gui/entry/EditEntryWidget.cpp]: The whole else block is completely redundant. Please consolidate to avoid code duplication.

File: src/gui/entry/EditEntryWidget.cpp (hunk 9 above)
Comment (TheZ3ro) [reply src/gui/entry/EditEntryWidget.cpp]: I've excluded the codestyle commit

File: src/autotype/AutoType.cpp [hunk 11]
```diff
@@ -665,3 +612,70 @@ bool AutoType::windowMatchesUrl(const QString& windowTitle, const QString& resol
...
+{
+    // checks things like {word 23}{F1 23}{~ 23}{% 23}{^}{F12}{(}{) 23}{[}{[}{]}{Delay=23}{+}{-}~+%@fixedstring
+    QString allowRepetition = "(\\s\\d*){0,1}";
+    QString normalCommands = "[A-Z:]*" + allowRepetition; // the ":" allows custom commands
+    QString specialLiterals = "[\\^\\%\\(\\)~\\{\\}\\[\\]\\+-]" + allowRepetition;
+    QString functionKeys = "(F[1-9]" + allowRepetition + "|F1[0-2])" + allowRepetition;
+    QString numpad = "NUMPAD\\d" + allowRepetition;
+    QString delay = "DELAY=\\d+";
+    QString beep = "BEEP\\s\\d*\\s\\d*";
+    QString vkey = "VKEY(-[EN]X){0,1}" + allowRepetition;
```
[the remark this answers is not in the document]
Comment (TheZ3ro) [reply src/autotype/AutoType.cpp]: ex. `{VKEY 5}`, vkey will match `VKEY` and allowRepetition will match ` 5`. The actual value is matched, am I wrong?

File: src/autotype/AutoType.cpp [hunk 12]
```diff
@@ -665,3 +612,70 @@ bool AutoType::windowMatchesUrl(const QString& windowTitle, const QString& resol
...
+    QRegExp autoTypeSyntax("(" + shortcutKeys + "|" + fixedStrings + "|\\{(" + normalCommands + "|" + specialLiterals +
+                           "|" + functionKeys + "|" + numpad + "|" + delay + "|" + beep + "|" + vkey + ")\\})*");
+    autoTypeSyntax.setCaseSensitivity(Qt::CaseInsensitive);
+    autoTypeSyntax.setPatternSyntax(QRegExp::RegExp);
+    return autoTypeSyntax.exactMatch(string);
+}
+
+bool AutoType::checkHighDelay(const QString& string)
+{
+    QRegExp highDelay("\\{DELAY\\s\\d{5,}\\}"); // 5 digit numbers(10 seconds) are too much
```
[the remark this answers is not in the document]
Comment (TheZ3ro) [reply src/autotype/AutoType.cpp]: This only check for `{DELAY X}` (the delay applied between 2 actions). Up to 10 seconds is fine.

`{DELAY=X}` is the delay applied after every keypress. 10 seconds for every keypress is too huge

I will add another check for such case

File: src/autotype/AutoType.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (phoerious) [reply src/autotype/AutoType.cpp]: Syantax is {VKEY X} where X is the virtual key code. See https://keepass.info/help/base/autotype.html and https://msdn.microsoft.com/en-us/library/windows/desktop/dd375731.aspx

Repetition would be {VKEY X N}

[the remark this answers is not in the document]
Comment (TheZ3ro) [reply src/autotype/AutoType.cpp]: Okok, allowRepetition checks for a digit but the virtual keycode is `VK_LBUTTON` or `0x01` so should fit in a `VKEY\s\w+`

[the remark this answers is not in the document]
Comment (phoerious) [reply src/autotype/AutoType.cpp]: That's how I understood it, yes. But anyway, even if the key code were a number, the old regex would only match a single number (or none), but needed to match two.

**Comment by TheZ3ro** [comment]
@phoerious should be ok now

File: tests/TestAutoType.cpp [hunk 13]
```diff
@@ -282,3 +282,30 @@ void TestAutoType::testGlobalAutoTypeRegExp()
...
+    QCOMPARE(true, AutoType::checkSyntax("{NUMPAD1 3}"));
+
+    QCOMPARE(true, AutoType::checkSyntax("{BEEP 3 3}"));
+    QCOMPARE(false, AutoType::checkSyntax("{BEEP 3}"));
+
+    QCOMPARE(true, AutoType::checkSyntax("{VKEY 0x01}"));
+    QCOMPARE(true, AutoType::checkSyntax("{VKEY VK_LBUTTON}"));
+    QCOMPARE(true, AutoType::checkSyntax("{VKEY-EX 0x01}"));
+    // Bad sequence
+    QCOMPARE(false, AutoType::checkSyntax("{{{}}{}{}}{{}}"));
```
Comment (weslly) [inline tests/TestAutoType.cpp]: @TheZ3ro While you are at it, would be nice to have `{{}` and `{}}` covered by these tests.

File: tests/TestAutoType.cpp [hunk 14]
```diff
@@ -282,3 +282,30 @@ void TestAutoType::testGlobalAutoTypeRegExp()
     QCOMPARE(m_test->actionChars(), QString("custom_attr_third"));
     m_test->clearActions();
 }
+
+void TestAutoType::testAutoTypeSyntaxChecks()
+{
+    // Huge sequence
+    QCOMPARE(true, AutoType::checkSyntax("{word 23}{F1 23}{~ 23}{% 23}{^}{F12}{(}{) 23}{[}{[}{]}{Delay=23}{+}{-}~+%@fixedstring"));
```
Comment (weslly) [inline tests/TestAutoType.cpp]: @TheZ3ro I just noticed there's no support for `{[}` and `{]}`, can you add it along with these changes or would  be better to open a separate PR?

Comment (weslly) [reply tests/TestAutoType.cpp]: Also, `{-}` should trigger a syntax error, since the [correct code is `{SUBSTRACT}`](https://keepass.info/help/base/autotype.html)

Comment (TheZ3ro) [reply tests/TestAutoType.cpp]: Why `{word}` should trigger an error? Because it's not a supported literal ?
Sadly RN we don't have a filter on supported literals so unless it has a different scheme (like `{BEEP X Y}`) we have to blindly accept it, then the action parser will skip unsupported ones

Comment (TheZ3ro) [reply tests/TestAutoType.cpp]: For the `{[}` literals better make a new PR, I will work on that after merging this old one

File: tests/TestAutoType.cpp (hunk 13 above)
Comment (TheZ3ro) [reply tests/TestAutoType.cpp]: Done

File: tests/TestAutoType.cpp (hunk 14 above)
Comment (weslly) [reply tests/TestAutoType.cpp]: >then the action parser will skip unsupported ones

Seems fair, since it won't break the entire sequence

File: src/gui/entry/EditEntryWidget.cpp [hunk 15]
```diff
@@ -31,7 +31,9 @@
 #include <QTemporaryFile>
 #include <QMimeData>
 #include <QEvent>
+#include <iostream>
```
Comment (phoerious) [inline src/gui/entry/EditEntryWidget.cpp]: Remove iostream include.

Comment (TheZ3ro) [reply src/gui/entry/EditEntryWidget.cpp]: Fixed/rebased
